src/docfx/lib/log/Reporter.cs

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Text.RegularExpressions;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Docs.Build
+{
+    internal sealed class Reporter : IDisposable
+    {
+        private readonly object _lock = new object();
+        private Lazy<TextWriter> _output;
+        private bool _stable;
+
+        public void OutputToFile(string outputFilePath, bool stable)
```

Comment:
not thread safe?

Reply:
agree

---
In reply to: [187504049](https://github.com/dotnet/docfx/pull/2758#discussion_r187504049) [](ancestors = 187504049)

Reply:
can it be a constructor?

---
In reply to: [187504180](https://github.com/dotnet/docfx/pull/2758#discussion_r187504180) [](ancestors = 187504180,187504049)

Reply:
This method is intended to be only be called once on a single instance, thus thread safe, see `Debug.Assert`.

It cannot be put into constructor because, before config is loaded, we don't know the output location so errors are written to console only, after that, errors will be logged to both console and file.

---
In reply to: [187504215](https://github.com/dotnet/docfx/pull/2758#discussion_r187504215) [](ancestors = 187504215,187504180,187504049)

Reply:
We could buffer logs before an output is specified and write them all to file once we have the output, provided we don't hold too much logs in memory.

---
In reply to: [187523738](https://github.com/dotnet/docfx/pull/2758#discussion_r187523738) [](ancestors = 187523738,187504215,187504180,187504049)